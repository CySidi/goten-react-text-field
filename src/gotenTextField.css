.error {
 border:1px solid red;
}

.error-message {
    color: red;
    font-size: 10px;
}

.label {
    font-size: 15px;
}