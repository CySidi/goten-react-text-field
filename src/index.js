import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { PubSub } from 'pubsub-js'

import './gotenTextField.css'


const defaultValue = ''
const defaultType = 'text'
const defaultIsRequiredMessage = '* is required'
const defaultTypeErrorMessage = '{label} must be a {type}'

const pubsubMessageResponse = '_RESPONSE'

const textInputState = {
    value: defaultValue,
    error: {
        error: false,
        errorMessage: ''
    }
}

const regex = {
    email: /^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$/,
    number: /^-?[0-9]+(.[0-9]+)?$/,
    date: /^[0-9]{4}-[0-9]{2}-[0-9]{2}$/,
    time: /^[0-9]{2}:[0-9]{2}$/,
    url: /^[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)$/,
    password: /./,
    text: /./
}

export class GotenTextField extends Component {
    constructor(props) {
        super(props)
        this.state = textInputState
        this.state.value = this.props.value ? this.props.value : defaultValue
        this.value = this.state.value
        this.type = this.props.type ? this.props.type : defaultType
        this.input = React.createRef()
        this.subscription = undefined
        this._formRegister()
    }

    _formRegister() {
        if (!this.subscription && this.props._pubsub_message)
            this.subscription = PubSub.subscribe(this.props._pubsub_message, _ => {
                PubSub.publish(this.props._pubsub_message + pubsubMessageResponse, !this.validate().error)
            })
    }

    componentWillUnmount() {
         if (this.subscription)
            PubSub.unsubscribe(this.subscription)
    }
    
    componentDidUpdate(props) {
        if (props.type !== this.props.type)
            this.type = this.props.type
    }

    render() {
        return (
            <div>
                { this.props.label && !this.props.componentLabel &&
                    <label className='label'>{this.props.label + ' '}</label>
                }
                { this.props.componentLabel }
                { this.props.multiline ?
                    React.cloneElement(<textarea />, this._getProps()) :
                    React.cloneElement(<input />, this._getProps())
                }
                <br/>
                { this.props.showError &&
                    <label className='error-message'>{this.state.error.errorMessage}</label>
                }
            </div>
        )
    }

    _getProps() {
        const auxProps = {
            ...this.props,
            className: this.props.className + (this.props.showError && this.state.error.error ? ' error' : ''),
            value: this.state.value,
            onChange: this._onChange,
            type: this.type,
            required: this.props.required,
            ref: this.input,
        }
        const { 
            bindContainer, 
            bindProp,
            componentLabel,
            errorMessage, 
            errorRequiredMessage,
            key,
            label,
            multiline,
            showError, 
            _pubsub_message, 
            ...props 
        } = auxProps
        return props
    }

    _onChange = (event) => {
        this._valueUpdate(event.target.value)
        if (this.props.onChange)
            this.props.onChange(event)
    }

    _valueUpdate(value) {
        new Promise((resolve) => {
            this.setState({ value })
            resolve()
        })
            .then(() => {
                this.value = this.state.value
                if (this.props.bindContainer !== undefined)
                    this.props.bindContainer[this.props.bindProp] = this.value
            })
    }

    validate() {
        const typeRegex = this.props.pattern ? new RegExp(this.props.pattern) : (regex[this.type] ? regex[this.type] : regex[defaultType])
        const defaultTypeSelectedErrorMessage = defaultTypeErrorMessage.replace('{type}', this.type).replace('{label}', this.props.label)
        let error = {
            error: false,
            errorMessage: ''  
        }
        if (this.props.required && this.value === defaultValue) {
            error = {
                error: true,
                errorMessage: this.props.errorRequiredMessage ? this.props.errorRequiredMessage : defaultIsRequiredMessage
            }
        } else if (!this.props.required && this.value === defaultValue) {
            null
        } else if (!typeRegex.test(this.value)) {
            error = {
                error: true,
                errorMessage: this.props.errorMessage ? this.props.errorMessage : defaultTypeSelectedErrorMessage
            }
        }
        this.setState({error})
        return error
    }
}

GotenTextField.propTypes = {
    bindContainer: PropTypes.object,
    bindProp: PropTypes.string,
    componentLabel: PropTypes.element,
    errorMessage: PropTypes.string,
    errorRequiredMessage: PropTypes.string,
    label: PropTypes.string,
    multiline: PropTypes.bool,
    pattern: PropTypes.string,
    placeholder: PropTypes.string,
    required: PropTypes.bool,
    showError: PropTypes.bool,
    type: PropTypes.string
}