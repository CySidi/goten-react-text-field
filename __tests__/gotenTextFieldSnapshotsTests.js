import React from 'react'
import renderer from 'react-test-renderer'

import { GotenTextField } from '../src'


describe('GotenTextField snapshots', () => {

    it('type text', () => {
        const type = 'text'
        const label = 'Text'
        const tree = renderer.create(
            <GotenTextField
                label={label}
                type={type}
            /> 
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
    
    it('type email', () => {
        const type = 'email'
        const label = 'Email'
        const tree = renderer.create(
            <GotenTextField
                label={label}
                type={type}
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
    
    it('type number', () => {
        const type = 'number'
        const label = 'Number'
        const tree = renderer.create(
            <GotenTextField
                label={label}
                type={type}
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
    
    it('type time', () => {
        const type = 'time'
        const label = 'Time'
        const tree = renderer.create(
            <GotenTextField
                label={label}
                type={type}
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
    
    it('type url', () => {
        const type = 'url'
        const label = 'Url'
        const tree = renderer.create(
            <GotenTextField
                label={label}
                type={type}
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
    
    it('type password', () => {
        const type = 'password'
        const label = 'Password'
        const tree = renderer.create(
            <GotenTextField
                label={label}
                type={type}
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
    
    it('type invalid', () => {
        const type = 'invalid'
        const label = 'Invalid'
        const tree = renderer.create(
            <GotenTextField
                label={label}
                type={type}
            />
        ).toJSON()
        expect(tree).toMatchSnapshot()
    })
})