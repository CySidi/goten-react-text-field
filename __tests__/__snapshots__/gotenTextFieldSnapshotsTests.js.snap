// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GotenTextField snapshots type email 1`] = `
<div>
  <label
    className="label"
  >
    Email 
  </label>
  <input
    className="undefined"
    label="Email"
    onChange={[Function]}
    type="email"
    value=""
  />
  <br />
</div>
`;

exports[`GotenTextField snapshots type invalid 1`] = `
<div>
  <label
    className="label"
  >
    Invalid 
  </label>
  <input
    className="undefined"
    label="Invalid"
    onChange={[Function]}
    type="invalid"
    value=""
  />
  <br />
</div>
`;

exports[`GotenTextField snapshots type number 1`] = `
<div>
  <label
    className="label"
  >
    Number 
  </label>
  <input
    className="undefined"
    label="Number"
    onChange={[Function]}
    type="number"
    value=""
  />
  <br />
</div>
`;

exports[`GotenTextField snapshots type password 1`] = `
<div>
  <label
    className="label"
  >
    Password 
  </label>
  <input
    className="undefined"
    label="Password"
    onChange={[Function]}
    type="password"
    value=""
  />
  <br />
</div>
`;

exports[`GotenTextField snapshots type text 1`] = `
<div>
  <label
    className="label"
  >
    Text 
  </label>
  <input
    className="undefined"
    label="Text"
    onChange={[Function]}
    type="text"
    value=""
  />
  <br />
</div>
`;

exports[`GotenTextField snapshots type time 1`] = `
<div>
  <label
    className="label"
  >
    Time 
  </label>
  <input
    className="undefined"
    label="Time"
    onChange={[Function]}
    type="time"
    value=""
  />
  <br />
</div>
`;

exports[`GotenTextField snapshots type url 1`] = `
<div>
  <label
    className="label"
  >
    Url 
  </label>
  <input
    className="undefined"
    label="Url"
    onChange={[Function]}
    type="url"
    value=""
  />
  <br />
</div>
`;
